README.txt
Hadoop EC2 Control Scripts
==========================

You can find the latest version of this documentation at:

http://www.cloudera.com/hadoop-ec2

Getting Started
===============

First, unpack the scripts on your system. For convenience, you may like to put
the top-level directory on your path.

You'll also need python (version 2.5 or newer) and the boto and simplejson
libraries. After you download boto and simplejson, you can install each in turn
by running the following in the directory where you unpacked the distribution:

% sudo python setup.py install

Alternatively, you might like to use the python-boto and python-simplejson RPM
and Debian packages.

You need to tell the scripts your AWS credentials. The simplest way to do this
is to set the environment variables (but see
http://code.google.com/p/boto/wiki/BotoConfig for other options):

    * AWS_ACCESS_KEY_ID - Your AWS Access Key ID
    * AWS_SECRET_ACCESS_KEY - Your AWS Secret Access Key

To configure the scripts, create a directory called .hadoop-ec2 (note the
leading ".") in your home directory. In it, create a file called
ec2-clusters.cfg with a section for each cluster you want to control. e.g.:

[my-hadoop-cluster]
ami=ami-6159bf08
instance_type=c1.medium
key_name=tom
availability_zone=us-east-1c
private_key=PATH_TO_PRIVATE_KEY
ssh_options=-i %(private_key)s -o StrictHostKeyChecking=no

The AMI chosen here is one with a Fedora OS. You may change this to be one of
the following:

AMI (bucket/name)                                                   ID
cloudera-ec2-hadoop-images/cloudera-hadoop-fedora-20090623-i386     ami-6159bf08
cloudera-ec2-hadoop-images/cloudera-hadoop-fedora-20090623-x86_64   ami-2359bf4a
cloudera-ec2-hadoop-images/cloudera-hadoop-ubuntu-20090623-i386     ami-ed59bf84
cloudera-ec2-hadoop-images/cloudera-hadoop-ubuntu-20090623-x86_64   ami-8759bfee

The architecture must be compatible with the instance type. For m1.small and
c1.medium instances use the i386 AMIs, while for m1.large, m1.xlarge, and
c1.xlarge instances use the x86_64 AMIs. One of the high CPU instances
(c1.medium or c1.xlarge) is recommended.

Then you can run the hadoop-ec2 script. It will display usage instructions when
invoked without arguments.

You can test that it can connect to AWS by typing:

% hadoop-ec2 list

LAUNCHING A CLUSTER
===================

To launch a cluster called "my-hadoop-cluster" with with 10 worker (slave) nodes
type:

% hadoop-ec2 launch-cluster my-hadoop-cluster 10

This will boot the master node and 10 worker nodes. When the nodes have started
and the Hadoop cluster has come up, the console will display a message like

  Browse the cluster at http://ec2-xxx-xxx-xxx-xxx.compute-1.amazonaws.com/

You can access Hadoop's web UI by visiting this URL. By default, port 80 is
opened for access from your client machine. You may change the firewall settings
(to allow access from a network, rather than just a single machine, for example)
by using the Amazon EC2 command line tools, or by using a tool like Elastic Fox.
The security group to change is the one named <cluster-name>-master.

For security reasons, traffic from the network your client is running on is
proxied through the master node of the cluster using an SSH tunnel (a SOCKS
proxy on port 6666). To set up the proxy run the following command:

% hadoop-ec2 proxy my-hadoop-cluster

Web browsers need to be configured to use this proxy too, so you can view pages
served by worker nodes in the cluster. The most convenient way to do this is to
use a proxy auto-config (PAC) file, such as this one:

  http://cloudera-public.s3.amazonaws.com/ec2/proxy.pac

If you are using Firefox, then you may find
FoxyProxy useful for managing PAC files. (If you use FoxyProxy, then you need to
get it to use the proxy for DNS lookups. To do this, go to Tools -> FoxyProxy ->
Options, and then under "Miscellaneous" in the bottom left, choose "Use SOCKS
proxy for DNS lookups".)

PERSISTENT CLUSTERS
===================

Hadoop clusters running on EC2 that use local EC2 storage (the default) will not
retain data once the cluster has been terminated. It is possible to use EBS for
persistent data, which allows a cluster to be shut down while it is not being
used.

Note: EBS support is a Beta feature.

First create a new section called "my-ebs-cluster" in the
.hadoop-ec2/ec2-clusters.cfg file.

Now we need to create storage for the new cluster. Create a temporary EBS volume
of size 100GiB, format it, and save it as a snapshot in S3. This way, we only
have to do the formatting once.

% hadoop-ec2 create-formatted-snapshot my-ebs-cluster 100

We create storage for a single master and for two slaves. The volumes to create
are described in a JSON spec file, which references the snapshot we just
created. Here is the contents of a JSON file, called
my-ebs-cluster-storage-spec.json:

{
  "master": [
    {
      "device": "/dev/sdj",
      "mount_point": "/ebs1",
      "size_gb": "100",
      "snapshot_id": "snap-268e704f"
    },
    {
      "device": "/dev/sdk",
      "mount_point": "/ebs2",
      "size_gb": "100",
      "snapshot_id": "snap-268e704f"
    }
  ],
  "slave": [
    {
      "device": "/dev/sdj",
      "mount_point": "/ebs1",
      "size_gb": "100",
      "snapshot_id": "snap-268e704f"
    },
    {
      "device": "/dev/sdk",
      "mount_point": "/ebs2",
      "size_gb": "100",
      "snapshot_id": "snap-268e704f"
    }
  ]
}


Each role (here "master" and "slave") is the key to an array of volume
specifications. In this example, the "slave" role has two devices ("/dev/sdj"
and "/dev/sdk") with different mount points, sizes, and generated from an EBS
snapshot. The snapshot is the formatted snapshot created earlier, so that the
volumes we create are pre-formatted. The size of the drives must match the size
of the snapshot created earlier.

Let's create actual volumes using this file.

% hadoop-ec2 create-storage my-ebs-cluster master 1 \
    my-ebs-cluster-storage-spec.json
% hadoop-ec2 create-storage my-ebs-cluster slave 2 \
    my-ebs-cluster-storage-spec.json

Now let's start the cluster with 2 slave nodes:

% hadoop-ec2 launch-cluster my-ebs-cluster 2

Login and run a job which creates some output.

% hadoop-ec2 login my-ebs-cluster

# hadoop fs -mkdir input
# hadoop fs -put /etc/hadoop/conf/*.xml input
# hadoop jar /usr/lib/hadoop/hadoop-*-examples.jar grep input output \
    'dfs[a-z.]+'

Look at the output:

# hadoop fs -cat output/part-00000 | head

Now let's shutdown the cluster.

% hadoop-ec2 terminate-cluster my-ebs-cluster

A little while later we restart the cluster and login.

% hadoop-ec2 launch-cluster my-ebs-cluster 2
% hadoop-ec2 login my-ebs-cluster

The output from the job we ran before should still be there:

# hadoop fs -cat output/part-00000 | head

RUNNING JOBS
============

When you launched the cluster, a hadoop-site.xml file was created in the
directory ~/.hadoop-ec2/<cluster-name>. You can use this to connect to the
cluster by setting the HADOOP_CONF_DIR enviroment variable (it is also possible
to set the configuration file to use by passing it as a -conf option to Hadoop
Tools):

% export HADOOP_CONF_DIR=~/.hadoop-ec2/my-hadoop-cluster

Let's try browsing HDFS:

% hadoop fs -conf -ls /

Running a job is straightforward:

% hadoop fs -mkdir input # create an input directory
% hadoop fs -conf -put $HADOOP_HOME/LICENSE.txt input # copy a file there
% hadoop jar $HADOOP_HOME/hadoop-*-examples.jar wordcount input output
% hadoop fs -cat output/part-00000 | head

Of course, these examples assume that you have installed Hadoop on your local
machine. It is also possible to launch jobs from within the cluster. First log
into the master node:

% hadoop-ec2 login my-hadoop-cluster

Then run a job as before:

# hadoop fs -mkdir input
# hadoop fs -put /etc/hadoop/conf/*.xml input
# hadoop jar /usr/lib/hadoop/hadoop-*-examples.jar grep input output 'dfs[a-z.]+'
# hadoop fs -cat output/part-00000 | head

TERMINATING A CLUSTER
=====================

When you've finished with your cluster you can stop it with the following
command.

NOTE: ALL DATA WILL BE LOST UNLESS YOU ARE USING EBS!

% hadoop-ec2 terminate-cluster my-hadoop-cluster

You can then delete the EC2 security groups with:

% hadoop-ec2 delete-cluster my-hadoop-cluster

AUTOMATIC CLUSTER SHUTDOWN
==========================

You may use the --auto-shutdown option to automatically terminate a cluster
a given time (specified in minutes) after launch. This is useful for short-lived
clusters where the jobs complete in a known amount of time.

If you want to cancel the automatic shutdown, then run

% hadoop-ec2 exec my-hadoop-cluster shutdown -c
% hadoop-ec2 update-slaves-file my-hadoop-cluster
% hadoop-ec2 exec my-hadoop-cluster /usr/lib/hadoop/bin/slaves.sh shutdown -c

TESTING PIG
===========

The scripts install Pig on the master node so you can issue queries in Pig Latin
from the master node. Here are a few pointers to get started with using it.

To run Pig, just type "pig". This will start the Grunt shell, connected to the
cluster. The following runs a simple Pig program against the sample data we
loaded into HDFS above:

grunt> A = LOAD 'input';
grunt> B = FILTER A BY $0 MATCHES '.*dfs[a-z.]+.*';
grunt> DUMP B;
grunt> quit

TESTING HIVE
============

The scripts also install Hive on the master node. To run it, type "hive".

hive> SHOW TABLES;

This should return successfully and show that there are no tables defined.
Run through the following commands to load and query some test data.

hive> CREATE TABLE test_table (foo INT, bar STRING);
hive> LOAD DATA LOCAL INPATH '/usr/share/doc/hive*/examples/files/kv1.txt*' OVERWRITE INTO TABLE test_table;
hive> SELECT MAX(foo) FROM test_table;
hive> DROP TABLE test_table;

CONFIGURATION NOTES
===================

It is possible to specify options on the command line: these take precedence
over any specified in the configuration file. For example:

% hadoop-ec2 launch-cluster --ami ami-2359bf4a --instance-type c1.xlarge \
  my-hadoop-cluster 10

This command launches a 10-node cluster using the specified AMI and instance
type, overriding the equivalent settings (if any) that are in the
"my-hadoop-cluster" section of the configuration file. Note that words in
options are separated by hyphens (--instance-type) while the corresponding
configuration parameter is are separated by underscores (instance_type).

The scripts install Hadoop RPMs or Debian packages (depending on the OS) at
instance boot time.

By default, CDH1 (http://archive.cloudera.com/docs/_cdh1_march_2009.html) is
installed (based on Hadoop 0.18.3). For this release, Hadoop configuration files
can be found in /etc/hadoop/conf and logs are in /var/log/hadoop.

You can also run other versions of Cloudera's Distribution for Hadoop. For
example the following uses the latest 0.20 version from the "testing" repository:

% hadoop-ec2 launch-cluster --env REPO=testing --env HADOOP_VERSION=0.20 \
  my-hadoop-cluster 10

CUSTOMIZATION
=============

You can specify a list of packages to install on every instance at boot time
using the --user-packages command-line option (or the user_packages
configuration parameter). Packages should be space-separated. Note that package
names should reflect the package manager being used to install them (yum or
apt-get depending on the OS).

Here's an example that installs RPMs for R and git:

% hadoop-ec2 launch-cluster --user-packages 'R git-core' my-hadoop-cluster 10

You have full control over the script that is run when each instance boots. The
default script, hadoop-ec2-init-remote.sh, may be used as a starting point to
add extra configuration or customization of the instance. Make a copy of the
script in your home directory, or somewhere similar, and set the
--user-data-file command-line option (or the user_data_file configuration
parameter) to point to the (modified) copy.  hadoop-ec2 will replace "%ENV%"
in your user data script with AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY,
USER_PACKAGES, AUTO_SHUTDOWN, and EBS_MAPPINGS, as well as extra parameters
supplied using the --env commandline flag.

Another way of customizing the instance, which may be more appropriate for
larger changes, is to create you own AMI using one of the base images listed in
the table above.

It's possible to use any AMI, as long as it i) runs (gzip compressed) user data
on boot, and ii) has Java installed.

RESOURCES
=========

    * Hadoop: http://hadoop.apache.org/
    * Pig: http://hadoop.apache.org/pig/
    * Hive: http://hadoop.apache.org/hive/
    * Hadoop on Amazon EC2: http://wiki.apache.org/hadoop/AmazonEC2
    * Cloudera's Distribution for Hadoop: http://www.cloudera.com/hadoop


CHANGES.txt
Hadoop EC2 Control Scripts Change Log

Trunk (unreleased)

  Create a unified PAC file for EC2 clusters.

  Enable map output compression by default.

  Add --auto-shutdown option for automatically shutting down instances after a
  given period (in minutes).

  Support both 0.18 and 0.20 clusters.

  Improve the proxy to return the PID of the proxy process. Typing

    eval `hadoop-ec2 proxy my-hadoop-cluster`

  starts the proxy and prints its PID. The PID is also set in the environment
  variable HADOOP_EC2_PROXY_PID.

  Add --client-cidr option which is used to set the firewall rules that allow
  the client to access the cluster.

  Add --force option to terminate-cluster and delete-storage.

  Fix bug in storage attachment code. (Contributed by Mark 'delduderino')

  Add support for --availability-zone option to create-storage.

0.2.0-beta

  Allow options to be specified on the command line, not just in a configuration
  file.

  Add --user-packages option for specifying user packages to install on
  instances.

  Unified remote instance boot script (hadoop-ec2-init-remote.sh) to support
  all combinations of OS (Fedora, Ubuntu) and storage (local disk, EBS).

  Make user_data_file configuration parameter optional.

  Make launch command block until Hadoop daemons are running.

  Generate local hadoop-site.xml to aid remote job submission.

  Use XFS filesystem for local disks.

0.1.1-beta

  Add authorization for port 80 access from client machine; and to master from
  itself. Also print master web server URL.

  Use standard Hadoop port numbers for namenode and jobtracker (8020, 8021).

  Generalized passing env variables to EC2 boot scripts.

  Add support for Ubuntu.

  Fix incorrect JSON spec in documentation.

0.1.0-beta

  First public release


hadoop/ec2/util.py
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import ConfigParser
import os
import socket
import urllib2

def bash_quote(text):
  """Quotes a string for bash, by using single quotes."""
  if text == None:
    return ""
  return "'%s'" % text.replace("'", "'\\''")

def bash_quote_env(env):
  """Quotes the value in an environment variable assignment."""
  if env.find("=") == -1:
    return env
  (var, value) = env.split("=")
  return "%s=%s" % (var, bash_quote(value))

def build_env_string(local_env_variables=[], env_strings=[], pairs={}):
  """Build a bash environment variable assignment"""
  env = ''
  if local_env_variables:
    for key in local_env_variables:
      if os.environ.has_key(key):
        env += "%s=%s " % (key, bash_quote(os.environ[key]))
  if env_strings:
    for env_string in env_strings:
      env += "%s " % bash_quote_env(env_string)
  if pairs:
    for key, val in pairs.items():
      env += "%s=%s " % (key, bash_quote(val))
  return env[:-1]

def merge_config_with_options(section_name, config, options):
  """
  Merge configuration options with a dictionary of options.
  Keys in the options dictionary take precedence.
  """
  d = {}
  try:
    for (key, value) in config.items(section_name):
      d[key] = value
  except ConfigParser.NoSectionError:
    pass
  for key in options:
    if options[key] != None:
      d[key] = options[key]
  return d

def url_get(url, timeout=10, retries=0):
  socket.setdefaulttimeout(timeout) # in Python 2.6 we can pass timeout to urllib2.urlopen
  attempts = 0
  while True:
    try:
      return urllib2.urlopen(url).read()
    except urllib2.URLError:
      attempts = attempts + 1
      if attempts > retries:
        raise

def xstr(s):
  """Sane string conversion: return an empty string if s is None."""
  return '' if s is None else str(s)


hadoop/ec2/cluster.py
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from boto.ec2.connection import EC2Connection
from boto.exception import EC2ResponseError
import logging
from hadoop.ec2.userdata import InstanceUserData
from hadoop.ec2.util import xstr
from subprocess import call;
import re;
import sys
import time

logger = logging.getLogger(__name__)

def get_clusters_with_role(role, state="running"):
  all = EC2Connection().get_all_instances()
  clusters = []
  for res in all:
    instance = res.instances[0];
    for group in res.groups:
      if group.id.endswith("-" + role) and instance.state == state:
        clusters.append(re.sub("-%s$" % re.escape(role), "", group.id))
  return clusters

class Cluster(object):
  """
  A cluster of EC2 instances. A cluster has a unique name.

  Instances running in the cluster run in a security group with the cluster's
  name, and also a name indicating the instance's role, e.g. <cluster-name>-foo
  to show a "foo" instance.
  """

  def __init__(self, name):
    self.name = name
    self.ec2Connection = EC2Connection()

  def get_ec2_connection(self):
    self.ec2Connection

  def get_cluster_group_name(self):
    return self.name

  def group_name_for_role(self, role):
    """
    Return the security group name for a instance in a given role.
    """
    return "%s-%s" % (self.name, role)

  def get_group_names(self, role):
    return [self.get_cluster_group_name(), self.group_name_for_role(role)]

  def _get_all_group_names(self):
    security_groups = self.ec2Connection.get_all_security_groups()
    security_group_names = [security_group.name for security_group in security_groups]
    return security_group_names

  def create_groups(self, role):
    """
    Create the security groups for a given role, including a group for the cluster
    if it doesn't exist.
    """
    security_group_names = self._get_all_group_names()

    cluster_group_name = self.get_cluster_group_name()
    if not cluster_group_name in security_group_names:
      self.ec2Connection.create_security_group(cluster_group_name, "Hadoop cluster (%s)" % (self.name))
      self.ec2Connection.authorize_security_group(cluster_group_name, cluster_group_name)
      # Allow SSH from anywhere
      self.ec2Connection.authorize_security_group(cluster_group_name, ip_protocol="tcp", from_port=22, to_port=22, cidr_ip="0.0.0.0/0")

    role_group_name = self.group_name_for_role(role)
    if not role_group_name in security_group_names:
      self.ec2Connection.create_security_group(role_group_name, "Hadoop %s (%s)" % (role, self.name))

  def authorize_role(self, role, from_port, to_port, cidr_ip):
    """
    Authorize access to machines in a given role from a given network.
    """
    role_group_name = self.group_name_for_role(role)
    # Revoke first to avoid InvalidPermission.Duplicate error
    self.ec2Connection.revoke_security_group(role_group_name, ip_protocol="tcp", from_port=from_port, to_port=to_port, cidr_ip=cidr_ip)
    self.ec2Connection.authorize_security_group(role_group_name, ip_protocol="tcp", from_port=from_port, to_port=to_port, cidr_ip=cidr_ip)

  def delete_groups(self, roles):
    """
    Delete the security groups for a given role, including the group for the cluster.
    """
    security_group_names = self._get_all_group_names()

    for role in roles:
      role_group_name = self.group_name_for_role(role)
      if role_group_name in security_group_names:
        self.ec2Connection.delete_security_group(role_group_name)
    cluster_group_name = self.get_cluster_group_name()
    if cluster_group_name in security_group_names:
      self.ec2Connection.delete_security_group(cluster_group_name)

  def get_instances(self, group_name, state_filter=None):
    """
    Get all the instances in a group, filtered by state.

    @param group_name: the name of the group
    @param state_filter: the state that the instance should be in (e.g. "running"),
                         or None for all states
    """
    all = self.ec2Connection.get_all_instances()
    instances = []
    for res in all:
      for group in res.groups:
        if group.id == group_name:
          for instance in res.instances:
            if state_filter == None or instance.state == state_filter:
              instances.append(instance)
    return instances

  def get_instances_in_role(self, role, state_filter=None):
    """
    Get all the instances in a role, filtered by state.

    @param role: the name of the role
    @param state_filter: the state that the instance should be in (e.g. "running"),
                         or None for all states
    """
    return self.get_instances(self.group_name_for_role(role), state_filter)

  def print_instance(self, role, instance):
    print "\t".join((role, instance.id,
      instance.image_id,
      "%-40s" % instance.dns_name,
      "%-24s" % instance.private_dns_name,
      instance.state, xstr(instance.key_name), instance.instance_type,
      str(instance.launch_time), instance.placement))

  def print_status(self, roles, state_filter="running"):
    for role in roles:
      for instance in self.get_instances_in_role(role, state_filter):
        self.print_instance(role, instance)

  def check_running(self, role, number):
    instances = self.get_instances_in_role(role, "running")
    if len(instances) != number:
      print "Expected %s instances in role %s, but was %s %s" % (number, role, len(instances), instances)
      return False
    else:
      return instances

  def launch_instances(self, role, number, image_id, key_name, user_data_file_template, replacements, instance_type='m1.small',
      placement=None):

    self.create_groups(role)
    user_data = InstanceUserData(user_data_file_template, replacements).read_as_gzip_stream()

    reservation = self.ec2Connection.run_instances(image_id, min_count=number, max_count=number, key_name=key_name,
      security_groups=self.get_group_names(role), user_data=user_data, instance_type=instance_type,
      placement=placement);
    return reservation

  def wait_for_instances(self, reservation):
    instances = [instance.id for instance in reservation.instances]
    # TODO(tom): should timeout
    while True:
      if self.all_started(self.ec2Connection.get_all_instances(instances)):
        break
      sys.stdout.write(".")
      sys.stdout.flush()
      time.sleep(1)

  def all_started(self, reservations):
    for res in reservations:
      for instance in res.instances:
        if instance.state != "running":
          return False
    return True

  def terminate(self):
    instances = self.get_instances(self.get_cluster_group_name(), "running")
    if instances:
      self.ec2Connection.terminate_instances([i.id for i in instances])


hadoop/ec2/userdata.py
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import with_statement

import gzip
import StringIO

class InstanceUserData(object):
  """
  The data passed to an EC2 instance on start up.
  """

  def __init__(self, filename, replacements={}):
    self.filename = filename
    self.replacements = replacements

  def read_file(self, filename):
    with open(filename, 'r') as f:
      return f.read()

  def read(self):
    contents = self.read_file(self.filename)
    for (match, replacement) in self.replacements.iteritems():
      if replacement == None:
        replacement = ''
      contents = contents.replace(match, replacement)
    return contents

  def read_as_gzip_stream(self):
    """
    Read and compress the data.
    """
    output = StringIO.StringIO()
    compressed = gzip.GzipFile(mode='wb', fileobj=output)
    compressed.write(self.read())
    compressed.close()
    return output.getvalue()


hadoop/ec2/commands.py
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""High-level commands that a user may want to run"""

from __future__ import with_statement

from hadoop.ec2.cluster import get_clusters_with_role
from hadoop.ec2.cluster import Cluster
from hadoop.ec2.storage import Storage
from hadoop.ec2.util import build_env_string
from hadoop.ec2.util import url_get
import logging
import os
import re
import socket
import sys
import time

logger = logging.getLogger(__name__)

ENV_WHITELIST = ('AWS_ACCESS_KEY_ID', 'AWS_SECRET_ACCESS_KEY')

MASTER = "master"
SLAVE = "slave"
ROLES = (MASTER, SLAVE)

DEFAULT_USER_DATA_FILE_TEMPLATE = os.path.join(sys.path[0], 'hadoop-ec2-init-remote.sh')

def list_all():
  """
  Find and print EC2 clusters that have a running 'master' instance
  """
  clusters = get_clusters_with_role(MASTER)
  if not clusters:
    print "No running clusters"
  else:
    for cluster in clusters:
      print cluster

def list(cluster_name):
  cluster=Cluster(cluster_name)
  cluster.print_status(ROLES)

def launch_master(cluster, image_id, key_name, user_data_file_template=None,
    instance_type='m1.small', placement=None, user_packages=None,
    auto_shutdown=None, env_strings=[], client_cidrs=[]):
  if user_data_file_template == None:
    user_data_file_template = DEFAULT_USER_DATA_FILE_TEMPLATE
  if cluster.check_running(MASTER, 0):
    return
  ebs_mappings=''
  storage = Storage(cluster)
  if storage.has_any_storage((MASTER,)):
    ebs_mappings = storage.get_mappings_string_for_role(MASTER)
  replacements = { "%ENV%": build_env_string(ENV_WHITELIST, env_strings, {
    "USER_PACKAGES": user_packages,
    "AUTO_SHUTDOWN": auto_shutdown,
    "EBS_MAPPINGS": ebs_mappings
  }) }
  reservation = cluster.launch_instances(MASTER, 1, image_id, key_name, user_data_file_template, replacements, instance_type, placement)
  print "Waiting for master to start (%s)" % str(reservation)
  cluster.wait_for_instances(reservation)
  print
  cluster.print_status((MASTER,))
  master = cluster.check_running(MASTER, 1)[0]
  _authorize_client_ports(cluster, master, client_cidrs)
  _create_client_hadoop_site_file(cluster, master)

def _authorize_client_ports(cluster, master, client_cidrs):
  if not client_cidrs:
    logger.debug("No client CIDRs specified, using local address.")
    client_ip = url_get('http://checkip.amazonaws.com/').strip()
    client_cidrs = ("%s/32" % client_ip,)
  logger.debug("Client CIDRs: %s", client_cidrs)
  for client_cidr in client_cidrs:
    # Allow access to port 80 on master from client
    cluster.authorize_role(MASTER, 80, 80, client_cidr)
    # Allow access to jobtracker UI on master from client (so we can see when the cluster is ready)
    cluster.authorize_role(MASTER, 50030, 50030, client_cidr)
    # Allow access to namenode and jobtracker via public address from master node
  master_ip = socket.gethostbyname(master.public_dns_name)
  cluster.authorize_role(MASTER, 8020, 8021, "%s/32" % master_ip)

def _create_client_hadoop_site_file(cluster, master):
  cluster_dir = os.path.join(os.environ['HOME'], '.hadoop-ec2/%s' % cluster.name)
  aws_access_key_id = os.environ['AWS_ACCESS_KEY_ID']
  aws_secret_access_key = os.environ['AWS_SECRET_ACCESS_KEY']
  if not os.path.exists(cluster_dir):
    os.makedirs(cluster_dir)
  with open(os.path.join(cluster_dir, 'hadoop-site.xml'), 'w') as f:
    f.write("""<?xml version="1.0"?>
<?xml-stylesheet type="text/xsl" href="configuration.xsl"?>
<!-- Put site-specific property overrides in this file. -->
<configuration>
<property>
  <name>hadoop.job.ugi</name>
  <value>root,root</value>
</property>
<property>
  <name>fs.default.name</name>
  <value>hdfs://%(master)s:8020/</value>
</property>
<property>
  <name>mapred.job.tracker</name>
  <value>%(master)s:8021</value>
</property>
<property>
  <name>hadoop.socks.server</name>
  <value>localhost:6666</value>
</property>
<property>
  <name>hadoop.rpc.socket.factory.class.default</name>
  <value>org.apache.hadoop.net.SocksSocketFactory</value>
</property>
<property>
  <name>fs.s3.awsAccessKeyId</name>
  <value>%(aws_access_key_id)s</value>
</property>
<property>
  <name>fs.s3.awsSecretAccessKey</name>
  <value>%(aws_secret_access_key)s</value>
</property>
<property>
  <name>fs.s3n.awsAccessKeyId</name>
  <value>%(aws_access_key_id)s</value>
</property>
<property>
  <name>fs.s3n.awsSecretAccessKey</name>
  <value>%(aws_secret_access_key)s</value>
</property>
</configuration>
""" % {'master': master.public_dns_name,
  'aws_access_key_id': aws_access_key_id,
  'aws_secret_access_key': aws_secret_access_key})

def launch_slaves(cluster, number, user_data_file_template=None,
    user_packages=None, auto_shutdown=None, env_strings=[]):
  if user_data_file_template == None:
    user_data_file_template = DEFAULT_USER_DATA_FILE_TEMPLATE
  instances = cluster.check_running(MASTER, 1)
  if not instances:
    return
  master = instances[0]
  ebs_mappings=''
  storage = Storage(cluster)
  if storage.has_any_storage((SLAVE,)):
    ebs_mappings = storage.get_mappings_string_for_role(SLAVE)
  replacements = { "%ENV%": build_env_string(ENV_WHITELIST, env_strings, {
    "USER_PACKAGES": user_packages,
    "AUTO_SHUTDOWN": auto_shutdown,
    "EBS_MAPPINGS": ebs_mappings,
    "MASTER_HOST": master.public_dns_name
  }) }
  reservation = cluster.launch_instances(SLAVE, number, master.image_id, master.key_name, user_data_file_template,
    replacements, master.instance_type, master.placement)
  print "Waiting for slaves to start"
  cluster.wait_for_instances(reservation)
  print
  cluster.print_status((SLAVE,))

def wait_for_hadoop(cluster, number):
  instances = cluster.check_running(MASTER, 1)
  if not instances:
    return
  master = instances[0]
  print "Waiting for jobtracker to start"
  previous_running = 0
  # TODO: timeout
  while True:
    try:
      actual_running = _number_of_tasktrackers(master.public_dns_name, 1)
      break
    except IOError:
      pass
    sys.stdout.write(".")
    sys.stdout.flush()
    time.sleep(1)
  print
  if number > 0:
    print "Waiting for %d tasktrackers to start" % number
    while actual_running < number:
      try:
        actual_running = _number_of_tasktrackers(master.public_dns_name, 5, 2)
        if actual_running != previous_running:
          sys.stdout.write("%d" % actual_running)
        sys.stdout.write(".")
        sys.stdout.flush()
        time.sleep(1)
        previous_running = actual_running
      except IOError:
        print
        print "Timeout waiting for jobtracker."
        return
    print

# The optional ?type=active is a difference between Hadoop 0.18 and 0.20
NUMBER_OF_TASK_TRACKERS=re.compile(
  r'<a href="machines.jsp(?:\?type=active)?">(\d+)</a>')

def _number_of_tasktrackers(jt_hostname, timeout, retries=0):
  jt_page = url_get("http://%s:50030/jobtracker.jsp" % jt_hostname, timeout, retries)
  m = NUMBER_OF_TASK_TRACKERS.search(jt_page)
  if m:
    return int(m.group(1))
  return 0

def print_master_url(cluster):
  instances = cluster.check_running(MASTER, 1)
  if not instances:
    return
  master = instances[0]
  print "Browse the cluster at http://%s/" % master.public_dns_name

def attach_storage(cluster, roles):
  storage = Storage(cluster)
  if storage.has_any_storage(roles):
    print "Waiting 10 seconds before attaching storage"
    time.sleep(10)
    for role in roles:
      storage.attach(role, cluster.get_instances_in_role(role, 'running'))
    storage.print_status(roles)
